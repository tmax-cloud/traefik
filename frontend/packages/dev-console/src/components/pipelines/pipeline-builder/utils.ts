import { history, resourcePathFromModel } from '@console/internal/components/utils';
import { apiVersionForModel } from '@console/internal/module/k8s';
import { ClusterTaskModel, PipelineModel } from '@console/internal/models';
import { Pipeline, PipelineResourceTask, PipelineResourceTaskParam, PipelineTask } from '../../../utils/pipeline-augment';
import { getTaskParameters } from '../resource-utils';
import { TASK_ERROR_STRINGS, TaskErrorType } from './const';
import { PipelineBuilderFormikValues, PipelineBuilderFormValues, TaskErrorMap } from './types';

export const getErrorMessage = (errorTypes: TaskErrorType[], errorMap: TaskErrorMap) => (taskName: string): string => {
  if (!taskName) {
    return TASK_ERROR_STRINGS[TaskErrorType.MISSING_NAME];
  }

  const errorList: TaskErrorType[] | undefined = errorMap?.[taskName];
  if (!errorList) return null;

  const hasRequestedError = errorList.filter(error => errorTypes.includes(error));
  return hasRequestedError.length > 0 ? TASK_ERROR_STRINGS[hasRequestedError[0]] : null;
};

export const taskParamIsRequired = (param: PipelineResourceTaskParam): boolean => {
  return false;
};

export const convertResourceToTask = (resource: PipelineResourceTask, runAfter?: string[]): PipelineTask => {
  return {
    name: resource.metadata.name,
    runAfter,
    taskRef: {
      kind: resource.kind === ClusterTaskModel.kind ? ClusterTaskModel.kind : undefined,
      name: resource.metadata.name,
    },
    params: getTaskParameters(resource).map(param => ({
      name: param.name,
      value: param.default,
      type: param.type
    })),
  };
};

export const getPipelineURL = (namespace: string) => {
  return `/k8s/ns/${namespace}/pipelines`;
};

const removeListRunAfters = (task: PipelineTask, listIds: string[]): PipelineTask => {
  if (task?.runAfter && listIds.length > 0) {
    // Trim out any runAfters pointing at list nodes
    const runAfter = (task.runAfter || []).filter(runAfterName => !listIds.includes(runAfterName));

    return {
      ...task,
      runAfter,
    };
  }

  return task;
};

const removeEmptyDefaultParams = (task: PipelineTask): PipelineTask => {
  if (task.params?.length > 0) {
    // Since we can submit, this param has a default; check for empty values and remove
    // 20210929 Policy changed
    task.params.forEach(param => { if (param.value === undefined) { param.value = (param.type && param.type === 'array') ?  [''] : ''; }; delete param.type; });
    return {
      ...task,
      //params: task.params.filter(param => !!param.value),
    };
  }

  return task;
};

export const convertBuilderFormToPipeline = (formValues: PipelineBuilderFormikValues, namespace: string, existingPipeline?: Pipeline): Pipeline => {
  const { name, resources, params, workspaces, tasks, listTasks, metadata } = formValues;
  const listIds = listTasks.map(listTask => listTask.name);
  const labels = metadata?.labels;

  return {
    ...existingPipeline,
    apiVersion: apiVersionForModel(PipelineModel),
    kind: PipelineModel.kind,
    metadata: {
      ...existingPipeline?.metadata,
      name,
      namespace,
      labels,
    },
    spec: {
      ...existingPipeline?.spec,
      params,
      resources,
      workspaces,
      tasks: tasks.map(task => removeEmptyDefaultParams(removeListRunAfters(task, listIds))),
    },
  };
};

export const convertPipelineToBuilderForm = (pipeline: Pipeline): PipelineBuilderFormValues => {
  if (!pipeline) return null;

  const {
    metadata: { name },
    spec: { params = [], resources = [], workspaces = [], tasks = [] },
  } = pipeline;

  return {
    name,
    params,
    resources,
    workspaces,
    tasks,
    listTasks: [],
  };
};

export const goToYAML = (existingPipeline?: Pipeline, namespace?: string) => {
  history.push(existingPipeline ? `${resourcePathFromModel(PipelineModel, existingPipeline?.metadata?.name, existingPipeline?.metadata?.namespace)}/yaml` : `${getPipelineURL(namespace)}/~new`);
};
