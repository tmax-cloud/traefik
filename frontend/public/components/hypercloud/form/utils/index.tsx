export * from './modal-launcher';
export * from './modal-list';
