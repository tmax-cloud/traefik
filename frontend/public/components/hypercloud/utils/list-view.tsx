import { useFormContext, useFieldArray } from 'react-hook-form';
import * as React from 'react';
import * as _ from 'lodash-es';
import { Button } from '@patternfly/react-core';
import { MinusCircleIcon, PlusCircleIcon } from '@patternfly/react-icons';

/**
 * hook-form으로 데이터를 관리하면서 Item을 추가/제거 할 수 있는 리스트뷰 컴포넌트이다.
 * @prop {string} name - hook form에 등록할 리스트뷰의 name.
 * @prop {object[]} defaultValues - 리스트뷰에 기본으로 들어갈 값을 설정해주는 속성이다. key/value로 이루어진 리스트뷰에 기본으로 두 개의 row item이 추가된 상태로 설정해주고 싶다면, [{key: 'AAA', value: 'aaa'}, {key: 'BBB', value: 'bbb'}] 로 설정해주면 된다.
 * @prop {object} defaultItem - 새로운 item 추가 시 해당 item에 넣어줄 기본값들에 대한 설정을 해주는 속성이다. (예: {key: 'defaultKey', value:'defaultValue'})
 * @prop {Function} itemRenderer - 리스트뷰 하나의 row item을 커스터마이징 할 수 있는 함수. 속성을 설정하지 않으면 기본으로 key/value로 이루어진 itemRenderer가 들어간다.
 * @prop {JSX.Element} headerFragment - 리스트뷰의 header부분을 커스터마이징 할 수 있는 속성. 속성을 설정하지 않으면 기본으로 KEY, VALUE로 구성된 헤더컴포넌트가 들어간다.
 * @prop {string} addButtonText - item 추가버튼의 텍스트. 기본값은 'Add'이다.
 */
export const ListView: React.FC<ListViewProps> = ({ name, methods, defaultItem = { key: '', value: '' }, itemRenderer, headerFragment, addButtonText, defaultValues, deleteButtonText, maxLength }) => {
  const methods_ = methods || useFormContext();
  const { control, register, getValues, setValue } = methods_;
  const { fields, append, remove } = useFieldArray({ control, name: name });

  const DefaultListHeaderFragment = (
    <div className="row pairs-list__heading">
      <div className="col-xs-4 text-secondary text-uppercase">KEY</div>
      <div className="col-xs-4 text-secondary text-uppercase">VALUE</div>
      <div className="col-xs-1 co-empty__header" />
    </div>
  );

  React.useEffect(() => {
    if (!!defaultValues) {
      // MEMO : name에 []이 들어가있으면 setValue 에러남. test[0].values 대신 test.0.values 형식으로 들어가있어야됨
      setValue(name, defaultValues);
    }
  }, [name]);

  const DefaultListItemRenderer = (register, name, item, index, ListActions, ListDefaultIcons) => {
    return (
      <div className="row" key={item.id} data-testid="row">
        <div className="col-xs-4 pairs-list__name-field">
          <input ref={register()} className="pf-c-form-control" name={`${name}[${index}].key`} data-testid={`${name}[${index}].key`} defaultValue={item.key}></input>
        </div>
        <div className="col-xs-4 pairs-list__value-field">
          <input ref={register()} className="pf-c-form-control" name={`${name}[${index}].value`} data-testid={`${name}[${index}].value`} defaultValue={item.value}></input>
        </div>
        <div className="col-xs-1 pairs-list__action">
          <Button
            type="button"
            data-test-id="pairs-list__delete-btn"
            className="pairs-list__span-btns"
            onClick={() => {
              ListActions.remove(index);
            }}
            variant="plain"
          >
            {ListDefaultIcons.deleteIcon}
          </Button>
        </div>
      </div>
    );
  };

  const deleteIcon = (
    <>
      <MinusCircleIcon className="pairs-list__side-btn pairs-list__delete-icon" />
      <span className="sr-only">Delete</span>
    </>
  );

  // MEMO : ListView컴포넌트 안에서 ref가 없는 다른 공통컴포넌트를 사용해 itemRenderer를 구성할 때, item remove할 때 그 안에 값들의 싱크가 맞지 않는 경우가 있어서,
  // 다시 원래 값으로 세팅해주는 함수. (임시 방안) 사용예시는 create-role.tsx 참고.
  const registerWithInitValue = (name, defaultValue) => {
    const isRegistered = _.get(getValues(), name);

    if (!isRegistered) {
      register(name);
      setValue(name, defaultValue);
    }
  };

  const ListActions = {
    append: append,
    remove: remove,
    getValues: getValues,
    registerWithInitValue: registerWithInitValue,
  };

  const ListDefaultIcons = {
    deleteIcon: deleteIcon,
  };

  const itemList = itemRenderer ? fields.map((item, index) => itemRenderer(methods_, name, item, index, ListActions, ListDefaultIcons, deleteButtonText)) : fields.map((item, index) => DefaultListItemRenderer(register, name, item, index, ListActions, ListDefaultIcons));
  const validLength = maxLength => {
    if (maxLength) {
      return maxLength > itemList.length;
    } else {
      return true;
    }
  };
  return (
    <div>
      {headerFragment ? headerFragment : DefaultListHeaderFragment}
      {itemList}
      {validLength(maxLength) && (
        <div className="row col-xs-12">
          <Button
            className="pf-m-link--align-left"
            data-test-id="pairs-list__add-btn"
            onClick={() => {
              append(defaultItem);
            }}
            type="button"
            variant="link"
          >
            <PlusCircleIcon data-test-id="pairs-list__add-icon" className="co-icon-space-r" />
            {!!addButtonText ? addButtonText : 'Add'}
          </Button>
        </div>
      )}
    </div>
  );
};

type ListViewProps = {
  name: string;
  defaultItem?: object;
  itemRenderer?: Function;
  headerFragment?: JSX.Element;
  deleteButtonText?: string;
  addButtonText?: string;
  methods?: any;
  maxLength?: number;
  defaultValues?: object[];
};
