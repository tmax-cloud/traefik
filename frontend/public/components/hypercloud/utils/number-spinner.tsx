import * as _ from 'lodash-es';
import * as React from 'react';
import * as classNames from 'classnames';
import { Button } from '@patternfly/react-core';
import { MinusSquareIcon, PlusSquareIcon } from '@patternfly/react-icons';
import { useFormContext } from 'react-hook-form';

export const NumberSpinner: React.FC<NumberSpinnerProps> = ({ className, initialValue, min, max, name, ...inputProps }) => {
  const { register } = useFormContext();
  const [value, setValue] = React.useState(initialValue);
  const changeValueBy = operation => {
    setValue(_.toInteger(value) + operation);
  };
  return (
    <div>
      <Button data-testid="minus-button" onClick={() => changeValueBy(-1)} type="button" variant="plain" isDisabled={!_.isNil(min) && value <= min} aria-label="Decrement" className="co-m-number-spinner__button">
        <MinusSquareIcon className="co-m-number-spinner__down-icon" />
      </Button>
      <input name={name} type="number" ref={register({ min: min, max: max })} value={value} onChange={(e: any) => setValue(e.target.value)} className={classNames(className, 'co-m-number-spinner__input', 'hc-number-spinner__input')} {...inputProps}></input>
      <Button data-testid="plus-button" onClick={() => changeValueBy(1)} type="button" variant="plain" isDisabled={!_.isNil(max) && value >= max} aria-label="Increment" className="co-m-number-spinner__button">
        <PlusSquareIcon className="co-m-number-spinner__up-icon" />
      </Button>
    </div>
  );
};

type NumberSpinnerProps = {
  className?: string;
  initialValue?: number;
  min?: number;
  max?: number;
  name?: string;
} & React.HTMLProps<HTMLInputElement>;

NumberSpinner.defaultProps = {
  name: 'numberSpinner',
  initialValue: 0,
};
