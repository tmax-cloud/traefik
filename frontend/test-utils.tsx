import * as React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from './jest-setup';

// MEMO : redux Provider로 감싼 구조로 @testing-library/react의 render함수를 래핑함.
const render = (ui: React.ReactElement, options?) => {
  const Wrapper = ({ children }: { children?: React.ReactNode }) => {
    return <Provider store={store}>{!!options?.wrapper ? options.wrapper({ children: children }) : children}</Provider>;
  };

  // MEMO : options로 들어온 wrapper는 Wrapper컴포넌트에서 사용하고, 여기서 wrapper속성을 Wrapper로 덮어씌워서 설정해 줌.
  return rtlRender(ui, { ...options, wrapper: Wrapper });
};

export * from '@testing-library/react';
export { render };
