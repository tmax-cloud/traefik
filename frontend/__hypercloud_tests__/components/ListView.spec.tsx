import * as React from 'react';
import { ListView } from '@console/internal/components/hypercloud/utils/list-view';
import { NumberSpinner } from '@console/internal/components/hypercloud/utils/number-spinner';
import { render, act } from '../../test-utils';
import userEvent from '@testing-library/user-event';
import { useForm, FormProvider } from 'react-hook-form';
import { Button } from '@patternfly/react-core';

const mockSubmit = jest.fn(data => {});

const renderListViewForm = (props, formDefaultValues?: object) => {
  return render(<ListView {...props} />, {
    wrapper: ({ children }) => {
      const methods = useForm(formDefaultValues ? { defaultValues: formDefaultValues } : {});
      return (
        <FormProvider {...methods}>
          <form
            onSubmit={methods.handleSubmit(data => {
              mockSubmit(data);
            })}
          >
            {children}
            <button type="submit">Submit</button>
          </form>
        </FormProvider>
      );
    },
  });
};

describe('ListView Test', () => {
  beforeEach(() => {
    // Empty
  });

  it('react-hook-form defaultValues 초기세팅 테스트입니다.', () => {
    const { getAllByTestId, getByTestId } = renderListViewForm(
      { name: 'metadata.listview' },
      {
        metadata: {
          listview: [
            { key: 'A', value: 'a' },
            { key: 'B', value: 'b' },
          ],
        },
      },
    );

    const keyArray = [];
    getAllByTestId('row').forEach((rowItem, index) => {
      keyArray.push({ key: (getByTestId(`metadata.listview[${index}].key`) as HTMLInputElement).value, value: (getByTestId(`metadata.listview[${index}].value`) as HTMLInputElement).value });
    });
    expect(keyArray).toEqual([
      { key: 'A', value: 'a' },
      { key: 'B', value: 'b' },
    ]);
  });

  it('addButtonText로 지정한 값이 추가버튼의 text로 설정되어야 합니다.', () => {
    const { getByText } = renderListViewForm({ name: 'metadata.listview', addButtonText: 'Add Key/Value' });
    getByText('Add Key/Value');
  });

  it('아이템 추가 될 때 name이 형식에 맞게 세팅되고, defaultValue로 지정한 값들이 기본으로 채워져야 합니다.', () => {
    const { getByText, getByTestId } = renderListViewForm({ name: 'metadata.listview', addButtonText: 'Add Key/Value', defaultItem: { key: 'defaultKey', value: 'defaultValue' } });
    userEvent.click(getByText('Add Key/Value'));
    expect(getByTestId('metadata.listview[0].key')).toHaveValue('defaultKey');
    expect(getByTestId('metadata.listview[0].value')).toHaveValue('defaultValue');
  });

  it('아이템 추가 후 리스트뷰 중간 아이템을 제거하면 각 row의 순서 및 값이 유지 되어야 합니다.', () => {
    const { container, getByText, getAllByTestId, getByTestId, getAllByText } = renderListViewForm({
      name: 'metadata.listview',
      addButtonText: 'Add Key/Value',
      defaultValues: [
        { key: 'AAA', value: 'aaa' },
        { key: 'BBB', value: 'bbb' },
        { key: 'CCC', value: 'ccc' },
      ],
    });

    userEvent.click(getByText('Add Key/Value'));
    userEvent.type(getByTestId('metadata.listview[3].key'), 'DDD');
    userEvent.type(getByTestId('metadata.listview[3].value'), 'ddd');

    userEvent.click(getAllByText('Delete')[1]);

    const keyArray = [];
    getAllByTestId('row').forEach((rowItem, index) => {
      keyArray.push((getByTestId(`metadata.listview[${index}].key`) as HTMLInputElement).value);
    });
    expect(keyArray).toEqual(['AAA', 'CCC', 'DDD']);
    expect(container).toMatchSnapshot();
  });

  it('CustomListView 스냅샷 및 react-hook-form 싱크(Submit value) 테스트입니다.', async () => {
    const listHeaderFragment = (
      <div className="row pairs-list__heading">
        <div className="col-xs-4 text-secondary text-uppercase">NAME</div>
        <div className="col-xs-4 text-secondary text-uppercase">NUM</div>
        <div className="col-xs-1 co-empty__header" />
      </div>
    );

    const listItemRenderer = (methods, name, item, index, ListActions, ListDefaultIcons) => (
      <div className="row" key={item.id}>
        <div className="col-xs-4 pairs-list__name-field">
          <input ref={methods.register()} className="pf-c-form-control" data-testid={`metadata.spinnerNumList[${index}].name`} name={`metadata.spinnerNumList[${index}].name`} defaultValue={item.name}></input>
        </div>
        <div className="col-xs-4 pairs-list__value-field">
          <NumberSpinner initialValue={item.number} min={-15} max={15} data-testid={`metadata.spinnerNumList[${index}].number`} name={`metadata.spinnerNumList[${index}].number`} />
        </div>
        <div className="col-xs-1 pairs-list__action">
          <Button
            type="button"
            data-testid="delete-button"
            className="pairs-list__span-btns"
            onClick={() => {
              ListActions.remove(index);
            }}
            variant="plain"
          >
            {ListDefaultIcons.deleteIcon}
          </Button>
        </div>
      </div>
    );
    const { container, getByText, getAllByTestId } = renderListViewForm(
      {
        name: 'metadata.spinnerNumList',
        addButtonText: 'Add NumberSpinner',
        defaultValues: [
          { name: 'row1', number: 1 },
          { name: 'row2', number: 2 },
        ],
        headerFragment: listHeaderFragment,
        itemRenderer: listItemRenderer,
        defaultItem: { name: 'defaultName', number: 0 },
      },
      {},
    );

    // MEMO : item 하나 추가한 뒤 name='row2'인 아이템 삭제 후 Submit
    userEvent.click(getByText('Add NumberSpinner'));
    userEvent.click(getAllByTestId('delete-button')[1]);
    expect(container).toMatchSnapshot();
    await act(async () => userEvent.click(getByText('Submit')));
    expect(mockSubmit).toHaveBeenCalledTimes(1);
    expect(mockSubmit).toHaveBeenCalledWith({
      metadata: {
        spinnerNumList: [
          { name: 'row1', number: '1' },
          { name: 'defaultName', number: '0' },
        ],
      },
    });
  });
});
