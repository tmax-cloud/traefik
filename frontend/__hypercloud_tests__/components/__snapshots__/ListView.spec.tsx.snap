// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView Test CustomListView 스냅샷 및 react-hook-form 싱크(Submit value) 테스트입니다. 1`] = `
<div>
  <form>
    <div>
      <div
        class="row pairs-list__heading"
      >
        <div
          class="col-xs-4 text-secondary text-uppercase"
        >
          NAME
        </div>
        <div
          class="col-xs-4 text-secondary text-uppercase"
        >
          NUM
        </div>
        <div
          class="col-xs-1 co-empty__header"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col-xs-4 pairs-list__name-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.spinnerNumList[0].name"
            name="metadata.spinnerNumList[0].name"
            value="row1"
          />
        </div>
        <div
          class="col-xs-4 pairs-list__value-field"
        >
          <div>
            <button
              aria-label="Decrement"
              class="pf-c-button pf-m-plain co-m-number-spinner__button"
              data-testid="minus-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="co-m-number-spinner__down-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM92 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H92z"
                  transform=""
                />
              </svg>
            </button>
            <input
              class="co-m-number-spinner__input hc-number-spinner__input"
              data-testid="metadata.spinnerNumList[0].number"
              name="metadata.spinnerNumList[0].number"
              type="number"
              value="1"
            />
            <button
              aria-label="Increment"
              class="pf-c-button pf-m-plain co-m-number-spinner__button"
              data-testid="plus-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="co-m-number-spinner__up-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                  transform=""
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="col-xs-1 pairs-list__action"
        >
          <button
            class="pf-c-button pf-m-plain pairs-list__span-btns"
            data-testid="delete-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pairs-list__side-btn pairs-list__delete-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                transform=""
              />
            </svg>
            <span
              class="sr-only"
            >
              Delete
            </span>
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-xs-4 pairs-list__name-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.spinnerNumList[1].name"
            name="metadata.spinnerNumList[1].name"
            value="defaultName"
          />
        </div>
        <div
          class="col-xs-4 pairs-list__value-field"
        >
          <div>
            <button
              aria-label="Decrement"
              class="pf-c-button pf-m-plain co-m-number-spinner__button"
              data-testid="minus-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="co-m-number-spinner__down-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM92 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H92z"
                  transform=""
                />
              </svg>
            </button>
            <input
              class="co-m-number-spinner__input hc-number-spinner__input"
              data-testid="metadata.spinnerNumList[1].number"
              name="metadata.spinnerNumList[1].number"
              type="number"
              value="0"
            />
            <button
              aria-label="Increment"
              class="pf-c-button pf-m-plain co-m-number-spinner__button"
              data-testid="plus-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="co-m-number-spinner__up-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                  transform=""
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="col-xs-1 pairs-list__action"
        >
          <button
            class="pf-c-button pf-m-plain pairs-list__span-btns"
            data-testid="delete-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pairs-list__side-btn pairs-list__delete-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                transform=""
              />
            </svg>
            <span
              class="sr-only"
            >
              Delete
            </span>
          </button>
        </div>
      </div>
      <div
        class="row col-xs-12"
      >
        <button
          class="pf-c-button pf-m-link pf-m-link--align-left"
          data-test-id="pairs-list__add-btn"
          type="button"
        >
          <span
            class="pf-c-button__text"
          >
            <svg
              aria-hidden="true"
              class="co-icon-space-r"
              data-test-id="pairs-list__add-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                transform=""
              />
            </svg>
            Add NumberSpinner
          </span>
        </button>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`ListView Test 아이템 추가 후 리스트뷰 중간 아이템을 제거하면 각 row의 순서 및 값이 유지 되어야 합니다. 1`] = `
<div>
  <form>
    <div>
      <div
        class="row pairs-list__heading"
      >
        <div
          class="col-xs-4 text-secondary text-uppercase"
        >
          KEY
        </div>
        <div
          class="col-xs-4 text-secondary text-uppercase"
        >
          VALUE
        </div>
        <div
          class="col-xs-1 co-empty__header"
        />
      </div>
      <div
        class="row"
        data-testid="row"
      >
        <div
          class="col-xs-4 pairs-list__name-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.listview[0].key"
            name="metadata.listview[0].key"
            value="AAA"
          />
        </div>
        <div
          class="col-xs-4 pairs-list__value-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.listview[0].value"
            name="metadata.listview[0].value"
            value="aaa"
          />
        </div>
        <div
          class="col-xs-1 pairs-list__action"
        >
          <button
            class="pf-c-button pf-m-plain pairs-list__span-btns"
            data-test-id="pairs-list__delete-btn"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pairs-list__side-btn pairs-list__delete-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                transform=""
              />
            </svg>
            <span
              class="sr-only"
            >
              Delete
            </span>
          </button>
        </div>
      </div>
      <div
        class="row"
        data-testid="row"
      >
        <div
          class="col-xs-4 pairs-list__name-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.listview[1].key"
            name="metadata.listview[1].key"
            value="CCC"
          />
        </div>
        <div
          class="col-xs-4 pairs-list__value-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.listview[1].value"
            name="metadata.listview[1].value"
            value="ccc"
          />
        </div>
        <div
          class="col-xs-1 pairs-list__action"
        >
          <button
            class="pf-c-button pf-m-plain pairs-list__span-btns"
            data-test-id="pairs-list__delete-btn"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pairs-list__side-btn pairs-list__delete-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                transform=""
              />
            </svg>
            <span
              class="sr-only"
            >
              Delete
            </span>
          </button>
        </div>
      </div>
      <div
        class="row"
        data-testid="row"
      >
        <div
          class="col-xs-4 pairs-list__name-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.listview[2].key"
            name="metadata.listview[2].key"
            value="DDD"
          />
        </div>
        <div
          class="col-xs-4 pairs-list__value-field"
        >
          <input
            class="pf-c-form-control"
            data-testid="metadata.listview[2].value"
            name="metadata.listview[2].value"
            value="ddd"
          />
        </div>
        <div
          class="col-xs-1 pairs-list__action"
        >
          <button
            class="pf-c-button pf-m-plain pairs-list__span-btns"
            data-test-id="pairs-list__delete-btn"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pairs-list__side-btn pairs-list__delete-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                transform=""
              />
            </svg>
            <span
              class="sr-only"
            >
              Delete
            </span>
          </button>
        </div>
      </div>
      <div
        class="row col-xs-12"
      >
        <button
          class="pf-c-button pf-m-link pf-m-link--align-left"
          data-test-id="pairs-list__add-btn"
          type="button"
        >
          <span
            class="pf-c-button__text"
          >
            <svg
              aria-hidden="true"
              class="co-icon-space-r"
              data-test-id="pairs-list__add-icon"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                transform=""
              />
            </svg>
            Add Key/Value
          </span>
        </button>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
